libnavigation-core/src/main/java/com/mapbox/navigation/core/routeoptions/MapboxRouteOptionsUpdater.kt

```diff
@@ -71,77 +73,121 @@ class MapboxRouteOptionsUpdater(
-                            it.addAll(
-                                routeOptions.waypointTargetsList()!!.subList(
-                                    index + 1,
-                                    coordinates.size
-                                )
-                            )
-                        }
-                    }
+                    getUpdatedWaypointTargetsList(
+                        routeOptions,
+                        coordinates.size - remainingWaypoints - 1
+                    )
+                )
+                .waypointIndicesList(
+                    getUpdatedWaypointIndicesList(
+                        routeOptions,
+                        coordinates.size - remainingWaypoints - 1
+                    )
                 )
         }
 
         return RouteOptionsUpdater.RouteOptionsResult.Success(optionsBuilder.build())
     }
+
+    private fun getUpdatedWaypointNamesList(
+        routeOptions: RouteOptions,
+        lastPassedWaypointIndex: Int
+    ): MutableList<String> {
+        val waypointNamesList = routeOptions.waypointNamesList()
+        if (waypointNamesList.isNullOrEmpty()) {
+            return mutableListOf()
+        }
+        return mutableListOf<String>().also { updatedWaypointNamesList ->
+            var updatedStartWaypointNamesIndex = 0
+            routeOptions.waypointIndicesList()?.forEachIndexed { indx, waypointIndex ->
+                if (waypointIndex <= lastPassedWaypointIndex) {
+                    updatedStartWaypointNamesIndex = indx
+                }
+            }
+            updatedWaypointNamesList.add(waypointNamesList[updatedStartWaypointNamesIndex])
+            updatedWaypointNamesList.addAll(
+                waypointNamesList.subList(
+                    updatedStartWaypointNamesIndex + 1,
+                    waypointNamesList.size
+                )
+            )
+        }
+    }
+
+    private fun getUpdatedWaypointTargetsList(
+        routeOptions: RouteOptions,
+        lastPassedWaypointIndex: Int
+    ): MutableList<Point> {
+        val waypointTargetsList = routeOptions.waypointTargetsList()
+        if (waypointTargetsList.isNullOrEmpty()) {
+            return mutableListOf()
+        }
+        return mutableListOf<Point>().also { updatedWaypointTargetsList ->
+            var updatedStartWaypointTargetsIndex = 0
+            routeOptions.waypointIndicesList()?.forEachIndexed { indx, waypointIndex ->
+                if (waypointIndex <= lastPassedWaypointIndex) {
+                    updatedStartWaypointTargetsIndex = indx
+                }
+            }
+            updatedWaypointTargetsList.add(waypointTargetsList[updatedStartWaypointTargetsIndex])
+            updatedWaypointTargetsList.addAll(
+                waypointTargetsList.subList(
+                    updatedStartWaypointTargetsIndex + 1,
+                    waypointTargetsList.size
+                )
+            )
+        }
+    }
```

Comment:
Can we extract some shared logic from these two functions? They look really similar.

Reply:
👍